securedrop/journalist_app/api.py

```diff
@@ -0,0 +1,221 @@
+
+    @api.route('/sources/<filesystem_id>/', methods=['GET'])
+    @token_required
+    def single_source(filesystem_id):
+        source = get_or_404(Source, filesystem_id, Source.filesystem_id)
+        return jsonify(source.to_json()), 200
+
+    @api.route('/sources/<filesystem_id>/add_star/', methods=['POST'])
+    @token_required
+    def add_star(filesystem_id):
+        source = get_or_404(Source, filesystem_id, Source.filesystem_id)
+        utils.make_star_true(source.filesystem_id)
+        db.session.commit()
+        return jsonify({'message': 'Star added'}), 201
+
+    @api.route('/sources/<filesystem_id>/remove_star/', methods=['DELETE'])
+    @token_required
+    def remove_star(filesystem_id):
+        source = get_or_404(Source, filesystem_id, Source.filesystem_id)
+        utils.make_star_false(source.filesystem_id)
+        db.session.commit()
+        return jsonify({'message': 'Star removed'}), 200
+
+    @api.route('/sources/<filesystem_id>/submissions/', methods=['GET',
+                                                                 'DELETE'])
+    @token_required
+    def all_source_submissions(filesystem_id):
+        if request.method == 'GET':
+            source = get_or_404(Source, filesystem_id, Source.filesystem_id)
+            return jsonify(
+                {'submissions': [submission.to_json() for
+                                 submission in source.submissions]}), 200
+        elif request.method == 'DELETE':
+            source = get_or_404(Source, filesystem_id, Source.filesystem_id)
+            utils.delete_collection(source.filesystem_id)
+            return jsonify({'message': 'Source and submissions deleted'}), 200
+
+    @api.route('/sources/<filesystem_id>/submissions/<int:submission_id>/download/',  # noqa
+               methods=['GET'])
+    @token_required
+    def download_submission(filesystem_id, submission_id):
+        source = get_or_404(Source, filesystem_id, Source.filesystem_id)
+        submission = get_or_404(Submission, submission_id)
+
+        # Mark as downloaded
+        submission.downloaded = True
+        db.session.commit()
+
+        return send_file(current_app.storage.path(source.filesystem_id,
+                                                  submission.filename),
+                         mimetype="application/pgp-encrypted",
+                         as_attachment=True)
+
+    @api.route('/sources/<filesystem_id>/submissions/<int:submission_id>/',
+               methods=['GET', 'DELETE'])
+    @token_required
+    def single_submission(filesystem_id, submission_id):
+        if request.method == 'GET':
+            submission = get_or_404(Submission, submission_id)
+            return jsonify(submission.to_json()), 200
+        elif request.method == 'DELETE':
+            submission = get_or_404(Submission, submission_id)
+            source = get_or_404(Source, filesystem_id, Source.filesystem_id)
```

Comment:
Ahh. yeah so, this is because I wanted to make this `get_or_404` function reusable for the both the (default) case where we're filtering by the primary key in the table and for the case where we are selecting by a particular column (e.g. `filesystem_id` or the Cool New Up-and-coming Column `journalist_designation_slug`). The function `get_or_404` takes an optional `column` kwarg such that I can optionally `filter` by it. If the column is not provided, then it assumes the `object_id` is a primary key: 

```
def get_or_404(model, object_id, column=''):
    if column:
        result = model.query.filter(column == object_id).first()
    else:
        result = model.query.get(object_id)
    if result is None:
        abort(404)
    return result
```

What I can do is be explicit, e.g. `source = get_or_404(Source, filesystem_id, column=Source.filesystem_id)`. What do you think?

Reply:
Yeah that might make more sense. I would have done it like this:

```python
def get_or_404(model, object_id, column=None):
    query = {column if column else 'id': object_id}
    result = model.query.filter_by(*query).one_or_none()
    if result is None:
        abort(404)
    return result
```

`filter_by` only takes kwargs and doesn't allow arbitrary SQLAlchemy query expressions. Also, I prefer `one_or_none` because that will raise an exception if there are more than one result which I think is what we want to prevent cases where we expect a single result so we don't accidentally return something we shouldn't.